Rubberduck.VBEEditor/SafeComWrappers/Office.Core/CommandBar.cs

```diff
@@ -164,5 +165,31 @@ public override int GetHashCode()
         {
             return IsWrappingNullReference ? 0 : HashCode.Compute(Type, Id, Index, IsBuiltIn, Target.Parent);
         }
+
+        private bool _isReleased;
+        public virtual void Release(bool final = false)
+        {
+            if (IsWrappingNullReference || _isReleased || !Marshal.IsComObject(Target))
+            {
+                _isReleased = true;
+                return;
+            }
+
+            try
+            {
+                if (final)
+                {
+                    Marshal.FinalReleaseComObject(Target);
```

Comment:
According to the documentation, the return should be `0` if it's successful. I think it's good idea to log (and maybe throw? I doubt) if the return from the `Marshal.FinalReleaseComObject` is non-zero.

Reply:
To be honest, I just used the commented out code from `SafeComWrapper`.

Reply:
@bclothier good idea. At *trace* level?

Reply:
Sounds good for trace level support. More information earlier the better as we can ease off how much input to justify low level COM data we need later down the track.

Reply:
Well.... I was considering the scenario where the `FinalReleaseComObject` returns a nonzero result... That is IMO, a fatal error because its basically saying we can't release it and probably should not be at this moment. So an Error level seems more apropos there.

That said, we ought to *trace* the non-final calls to `ReleaseComObject`. That would greatly aid us in tracking whether non-final calls allow us to gently put down the referenced COM object in appropriate order. We shouldn't be relying on the `FinalReleaseComObject` to do it for us - that's somehow akin to lobbing a nuclear bomb in your room to clean it.

We need a pattern here.

IMO, `_released` ought be an `int` with default of `null` and take the returns from both `ReleaseComObject` and `FinalReleaseComObject` so that we can then expose a property `bool Released => ( _released == 0 )` which would be more truthful.

I hesitate to suggest but we might consider whether we should require a finalizer so that we can check the `Released`, and if not zero, effect a `FinalReleaseComObject` with a log at *warning* level indicating that we just employed the nuclear option.

Reply:
Before I push and, thus, make this comment thread stale, I would like to comment on the idea with the finalizer. I think it would not really help. When this class gets GCed, probably the RCW it wrapps does, too. In that case, there is no use in prolonging the life of the object and the RCW. Moreover, The CLR might hand back the very same RCW to multiple instances of a wrapper class. In that case, we must not final release at the point the wrapper gets finalized.

Reply:
@MDoerner, good point regarding not calling the `FinalReleaseComObject` inside a finalizer.  I agree. That said, I still think we should log whenever a class has been finalized without the `_rcwReferenceCount` being zero. Otherwise, the log will not show any missed classes and unless we implement the dispose pattern, the finalizer is only place where we ought test whether its zero and issue a warning when it is not.